include/alp/reference/blas1.hpp

```diff
@@ -2847,28 +2847,85 @@ namespace alp {
 	}
 
 	/**
-	 * Sort vectors, function available to user, e.g. to sort eigenvectors
+	 * Compute the permutation vector needed to sort the input vector according to
+	 * the provided \a cmp function.
 	 *
 	 * @param[in] toSort vector of indices to sort, should not be modified
 	 * @param[in] cmp function with strict weak ordering relation between indices, eg bool cmp(const Type1 &a, const Type2 &b)
 	 *            cmp must not modify the objects passed to it
 	 *
 	 * @param[out] permutation iterator over index permutations which sort toSort vector
 	 *
-	 * Complexity should be lower than O(n*log(n)), and space complexity should be lower than \Theta(n+T+P)
+	 * Complexity should be \Theta(n*log(n)), and space complexity should be \Theta(n+T+P)
 	 */
 	template<
-		typename IndexType, typename IndexStructure, typename IndexView, typename IndexImfR, typename IndexImfC,
+		typename IndexStructure, typename IndexView, typename IndexImfR, typename IndexImfC,
+		typename ValueType, typename ValueStructure, typename ValueView, typename ValueImfR, typename ValueImfC
+	>
+	RC sort(
+		Vector< size_t, IndexStructure, Density::Dense, IndexView, IndexImfR, IndexImfC, reference > &permutation,
+		const Vector< ValueType, ValueStructure, Density::Dense, ValueView, ValueImfR, ValueImfC, reference > &toSort
+	) noexcept {
+
+		internal::setInitialized( permutation, internal::getInitialized( toSort ) );
+
+		if( !internal::getInitialized( toSort ) ) {
+			return SUCCESS;
+		}
+
+		RC rc = alp::eWiseLambda(
+			[ ]( const size_t i, size_t &val ) {
+				val = i;
+			},
+			permutation
+		);
+
+		typedef Vector< size_t, IndexStructure, Density::Dense, IndexView, IndexImfR, IndexImfC, reference > PermType;
+
+		typename PermType::iterator it_begin = internal::begin( permutation );
+		typename PermType::iterator it_end = internal::end( permutation );
+
+		std::sort( it_begin, it_end, [ &toSort ]( const size_t& a, const size_t& b ) {
+			return toSort[ a ] < toSort[ b ];
+		});
+		
+		return rc;
+	}
+
+	template<
+		typename IndexStructure, typename IndexView, typename IndexImfR, typename IndexImfC,
 		typename ValueType, typename ValueStructure, typename ValueView, typename ValueImfR, typename ValueImfC,
-		typename Compare
+		typename CompareType
 	>
 	RC sort(
-		Vector< IndexType, IndexStructure, Density::Dense, IndexView, IndexImfR, IndexImfC, reference > &permutation,
+		Vector< size_t, IndexStructure, Density::Dense, IndexView, IndexImfR, IndexImfC, reference > &permutation,
 		const Vector< ValueType, ValueStructure, Density::Dense, ValueView, ValueImfR, ValueImfC, reference > &toSort,
-		Compare cmp
-		//PHASE &phase = EXECUTE
+		CompareType cmp
```

Comment:
can there be one version of `sort` with default `cmp = less_then` ?

Reply:
This will be addressed soon by allowing a single variant of the function which always expects a poset/total order.